File: DynamicPrgramming/730. Count Different Palindromic Subsequences/Solution.py
class Solution(object):
    def countPalindromicSubsequences(self, S):
        """
        :type S: str
        :rtype: int
        """
        n = len(S)
        f = [[0 for j in range(n)] for i in range(n)]
        MOD = 1e9+7
        for i in range(n-1, -1, -1):
            for j in range(i, n):
                if i==j:
                    f[i][j]=1
                elif j-i==1:
                    f[i][j]=2
                else:
                    if S[i]==S[j]:
                        left, right=i+1, j-1
                        while left<=right and S[left]!=S[i]:
                            left+=1
                        while left<=right and S[right]!=S[i]:
                            right-=1
                        if left>right:
                            #no same char found
                            f[i][j] = 2*f[i+1][j-1]+2
                        elif left==right:
                            #one same char found
                            f[i][j] = 2*f[i+1][j-1]+1
                        else:
                            #at least two char found
                            f[i][j] = 2*f[i+1][j-1]-f[left+1][right-1]
                    else:
                        f[i][j]=f[i][j-1]+f[i+1][j]-f[i+1][j-1]
                if f[i][j]<0:
                    f[i][j]+=MOD
                f[i][j]%=MOD
        return int(f[0][n-1])
        

File: DynamicPrgramming/730. Count Different Palindromic Subsequences/Solution.java
class Solution {
    static final long MOD = (long)(1e9+7);
    public int countPalindromicSubsequences(String S) {
        int n = S.length();
        long[][] f = new long[n][n];
        for (int i=n-1; i>=0; i--){
            for (int j=i; j<n; j++){
                if (i==j){
                    f[i][j]=1;
                }else if (j-i==1){
                    f[i][j]=2;
                }else{
                    if (S.charAt(i)==S.charAt(j)){
                        int left=i+1;
                        int right=j-1;
                        while (left<=right && S.charAt(left)!=S.charAt(i)){
                            left++;
                        }
                        while (left<=right && S.charAt(right)!=S.charAt(i)){
                            right--;
                        }
                        if (left>right){
                            f[i][j]=2*f[i+1][j-1]+2;
                        }else if (left==right){
                            f[i][j]=2*f[i+1][j-1]+1;
                        }else{
                            f[i][j]=2*f[i+1][j-1]-f[left+1][right-1];
                        }
                    }else{
                        f[i][j]=f[i][j-1]+f[i+1][j]-f[i+1][j-1];
                    }
                }
                if (f[i][j]<0){
                    f[i][j]+=MOD;
                }
                f[i][j]%=MOD;
            }
        }
        return (int)f[0][n-1];
    }
}